### PR TITLE
[QTL] Move kafka-extraction-namespace to the Lookup framework.

### PR DESCRIPTION
**Pull request by drcrallen**
The major changes to the PR are as follows:

The kafka extraction namespace stuff now works off the lookup extractor factory framework. It still uses the caching mechanisms in the main lookup extension, but uses some random UUID in the cache instead of trying to keep some sort of relationship to the lookup name (which is enforced by the Lookup framework in core).

This means the kafka-extraction-namespace must go through the main lookup framework.

### CODE REVIEW

**Comment by drcrallen** [comment]
```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.575 sec <<< FAILURE! - in io.druid.query.lookup.TestKafkaExtractionCluster
testSimpleRename(io.druid.query.lookup.TestKafkaExtractionCluster)  Time elapsed: 4.57 sec  <<< ERROR!
com.fasterxml.jackson.databind.JsonMappingException: Could not resolve type id 'kafka' into a subtype of [simple type, class io.druid.query.lookup.LookupExtractorFactory]
 at [Source: {"type":"kafka","kafkaTopic":"testTopic","kafkaProperties":{"request.required.acks":"1","zookeeper.connect":"127.0.0.1:54849/kafka","zookeeper.sync.time.ms":"200","zookeeper.session.timeout.ms":"10000"}}; line: 1, column: 2]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
	at com.fasterxml.jackson.databind.DeserializationContext.unknownTypeException(DeserializationContext.java:862)
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._findDeserializer(TypeDeserializerBase.java:167)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:99)
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:84)
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:132)
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:41)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3066)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2161)
	at io.druid.query.lookup.TestKafkaExtractionCluster.setUp(TestKafkaExtractionCluster.java:266)
```
actual problem, fixing

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    @Override
+    public String fromBytes(byte[] bytes)
+    {
+      return StringUtils.fromUtf8(bytes);
+    }
+  };
+
+  private final Object startStopLock = new Object();
+  private final ListeningExecutorService executorService;
+  private final AtomicLong doubleEventCount = new AtomicLong(0L);
```
Comment (nishantmonu51) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: why do we maintain doubleEventCount instead of eventCount ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+                final String message = messageAndMetadata.message();
+                if (key == null || message == null) {
+                  LOG.error("Bad key/message from topic [%s]: [%s]", topic, messageAndMetadata);
+                  continue;
+                }
+                doubleEventCount.incrementAndGet();
+                map.put(key, message);
+                doubleEventCount.incrementAndGet();
+                LOG.trace("Placed key[%s] val[%s]", key, message);
+              }
```
Comment (nishantmonu51) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: Any exception will lead to the runnable to exit and no more updates will be processed, do we need to catch any exception here ?

**Comment by nishantmonu51** [comment]
could you summarize all the changes in this PR in the description ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 1 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: I was trying to have a simple way to minimize race conditions and locking. I could do a read/write lock if this isn't good enough.

Basically, before and after the critical section the count is increased (as opposed to just doing before or just doing after).

The crux revolves around "What was the state of the map that produced this result" for computing cache key. Which, for a continuously mutable map is a little tricky.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 2 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: Probably yes, I'll look and see what is best to capture here.

File: docs/content/development/extensions-core/kafka-extraction-namespace.md [hunk 3]
```diff
@@ -2,52 +2,34 @@
...
-## Configuration
-
-The following options are used to define the behavior and should be included wherever the extension is included (all query servicing nodes):
-
-|Property|Description|Default|
-|--------|-----------|-------|
-|`druid.query.rename.kafka.properties`|A json map of kafka consumer properties. See below for special properties.|See below|
-
-The following are the handling for kafka consumer properties in `druid.query.rename.kafka.properties`
+The consumer properties `group.id` and `auto.offset.reset` CANNOT be set in `kafkaProperties` as they are set by the extension as `UUID.randomUUID().toString()` and `smallest` respectively.
```
Comment (b-slim) [inline docs/content/development/extensions-core/kafka-extraction-namespace.md]: i am curious what is the implication of this constraint ?

Comment (drcrallen) [inline docs/content/development/extensions-core/kafka-extraction-namespace.md]: auto.offeset.reset as smallest means "read all the data available in the topic" otherwise two different servers could replay different changelogs.

group.id means every instance is a unique consumer, so they should be accounted for as different consumers.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+  @JsonProperty
+  private final String kafkaTopic;
+
+  @JsonProperty
+  private final Map<String, String> kafkaProperties;
+
+  @JsonCreator
+  public KafkaLookupExtractorFactory(
+      @JacksonInject NamespaceExtractionCacheManager cacheManager,
+      @NotNull @JsonProperty(value = "kafkaTopic", required = true) final String kafkaTopic,
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: can we get ride of `NotNull` ?

Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: i don't think `required` does have any effect.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    return kafkaProperties;
+  }
+
+  @Override
+  public boolean start()
+  {
+    synchronized (startStopLock) {
+      if (started.get()) {
+        LOG.warn("Already started, not starting again");
+        return false;
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: you might want return `true` if it is started.

Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]:
```
@return true if start successfully started the {@link LookupExtractor}
```

as per javadoc.

Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: other places in druid idempotent *stop* is common, but not required. idempotent *start* is rare.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 2 above)
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: i agree with @nishantmonu51, you might wanna swallow some of the exceptions if not all.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+  // Overriden in tests
+  ConsumerConnector buildConnector(Properties properties)
+  {
+    return new kafka.javaapi.consumer.ZookeeperConsumerConnector(
+        new ConsumerConfig(properties)
+    );
+  }
+
+  ;
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: clean.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+  @Override
+  public boolean close()
+  {
+    synchronized (startStopLock) {
+      if (!started.get() || executorService.isShutdown()) {
+        LOG.info("Already shutdown, ignoring");
+        return false;
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: calling stop multiple times should return `true` thought

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+      }
+      started.set(false);
+      executorService.shutdownNow();
+      final ListenableFuture<?> future = this.future;
+      if (future != null) {
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: shall we do this before `executorService.shutdownNow();`

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+  }
+
+  @Override
+  public boolean replaces(@Nullable LookupExtractorFactory other)
+  {
+    if (this == other) {
+      return false;
+    }
+
+    if (other == null || getClass() != other.getClass()) {
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: what's the point of replacing with null ? that will raise NPEs for no reasons

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+
+    return !(getKafkaTopic().equals(that.getKafkaTopic()) && getKafkaProperties().equals(that.getKafkaProperties()));
+  }
+
+  @Override
+  public LookupExtractor get()
+  {
+    final Map<String, String> map = Preconditions.checkNotNull(mapRef.get(), "Not started");
+    final long startCount = doubleEventCount.get();
+    return new MapLookupExtractor(map, false)
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: you might wanna allow user to select if it is one to one ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 5 above)
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: if we call start twice, do you consider the second as failure ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 4 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: it does not

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 5 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: In most cases yes, From the API it does not say "if the lookup extractor is started" it says "if start successfully started"

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 8 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: shutdownNow *should* attempt to do a future.cancel(true). This little block is more for ensuring logging of an extreme failure case where the future isn't dying.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 10 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: fixed

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 9 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: fixed

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 7 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: fixed

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 6 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: fixed

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 4 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: removed

Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: you hate `NotNull` huh? :)

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 2 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: Fixed, please have a look at the change.

**Comment by drcrallen** [comment]
Merge problems from introspection PR, fixing

**Comment by drcrallen** [comment]
@b-slim I'm trying to hammer out some tests to prevent racy-ness, but otherwise this should be done.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+  };
+
+  private final Object startStopLock = new Object();
+  private final ListeningExecutorService executorService;
+  private final AtomicLong doubleEventCount = new AtomicLong(0L);
+  private final NamespaceExtractionCacheManager cacheManager;
+  private final String factoryId = UUID.randomUUID().toString();
+  private final AtomicReference<Map<String, String>> mapRef = new AtomicReference<>(null);
+
+  private AtomicBoolean started = new AtomicBoolean(false);
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: can we use this for startStop lock as well ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+  private volatile ListenableFuture<?> future = null;
+
+  @JsonProperty
+  private final String kafkaTopic;
+
+  @JsonProperty
+  private final Map<String, String> kafkaProperties;
+
+  @JsonProperty
+  private final long connectTimeout;
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: wondering if this is not part of the kafka properties ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 5 above)
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: i still think this should return `started.get()` which is true.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 13]
```diff
@@ -0,0 +1,423 @@
...
+            public void onSuccess(Object result)
+            {
+              LOG.debug("Success listening to [%s]", topic);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+              if (t instanceof CancellationException) {
+                LOG.debug("Topic [%s] cancelled", topic);
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: i guess this should be a warn ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+  {
+    final Map<String, String> map = Preconditions.checkNotNull(mapRef.get(), "Not started");
+    final long startCount = doubleEventCount.get();
+    return new MapLookupExtractor(map, isOneToOne())
+    {
+      @Override
+      public byte[] getCacheKey()
+      {
+        final byte[] idutf8 = StringUtils.toUtf8(factoryId);
+        if (startCount == doubleEventCount.get()) {
```
Comment (b-slim) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: can you please explain what this condition means ?

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 13 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: During normal shutdown the thread is interrupted and has the future's cancel flag set to true. I can put it as info if you want but warning is not correct.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 14 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: added

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 11 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: that is possible yes, changed

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 12 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: This is an extra safety valve above and beyond just the raw connection properties. If for some reason the setup process to connect to kafka is bugging out (in this class) this makes it so that the startup will fail rather than potentially loop forever.

File: extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java (hunk 5 above)
Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: that needs clarified in the API javadoc

Comment (drcrallen) [inline extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java]: @b-slim https://github.com/druid-io/druid/issues/2888

**Comment by b-slim** [comment]
@drcrallen can you explain how the caching mechanism works ?
I guess First the user need to know how caching works and where the cache is leaving.
Plus i am not sure how eviction is suppose to work here ? this impl is going to fill the cache with all the entry thought.

**Comment by drcrallen** [comment]
@b-slim this impl fills the cache with all entries without eviction.

**Comment by drcrallen** [comment]
@b-slim this does not change caching mechanism from prior impl I don't understand why caching would be a blocker for this PR

**Comment by b-slim** [comment]
@drcrallen i know that, my comments it to help the user understand how thing works by adding it to the docs. IMHO it is a serious limitation to mention.

**Comment by b-slim** [comment]
@drcrallen i am not blocking it, just update the DOCs to reflect the limitation.

**Comment by b-slim** [comment]
:+1: after Docs.

**Comment by drcrallen** [comment]
Will add docs very shortly

**Comment by drcrallen** [comment]
@b-slim added

```
# Limitations

Currently the Kafka lookup extractor feeds the entire kafka stream into a local cache. If you are using OnHeap caching, this can easily clobber your java heap if the kafka stream spews a lot of unique keys.
OffHeap caching should alleviate these concerns, but there is still a limit to the quantity of data that can be stored.
There is currently no eviction policy.
```

**Comment by drcrallen** [comment]
https://travis-ci.org/druid-io/druid/jobs/126506259 died in 

```

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running io.druid.indexing.kafka.KafkaDataSourceMetadataTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.022 sec - in io.druid.indexing.kafka.KafkaDataSourceMetadataTest
Running io.druid.indexing.kafka.KafkaIndexTaskTest
```

**Comment by nishantmonu51** [comment]
👍 , LGTM
